src/auth.rs

```diff
@@ -0,0 +1,135 @@
+use std::error::Error;
+use std::ops::Deref;
+
+use axum::extract::{Extension, FromRequest, Query, RequestParts};
+use axum::headers;
+use axum::http::header::SET_COOKIE;
+use axum::http::{HeaderMap, StatusCode};
+use axum::response::{IntoResponse, Redirect, Response};
+use axum::{async_trait, TypedHeader};
+use log::{debug, error, trace};
+use openidconnect::core::{CoreClient, CoreProviderMetadata, CoreResponseType, CoreUserInfoClaims};
+use openidconnect::ureq::http_client;
+use openidconnect::url::Url;
+use openidconnect::{
+    AccessToken, AuthenticationFlow, AuthorizationCode, ClientId, ClientSecret, CsrfToken,
+    IssuerUrl, Nonce, OAuth2TokenResponse, RedirectUrl, Scope,
+};
+use serde::Deserialize;
+
+const COOKIE_NAME: &str = "SESSION";
+
+#[derive(Debug, Deserialize)]
+pub struct AuthRequest {
+    pub code: String,
+    pub state: String,
+}
+
+pub fn openid_client(
+    redirect_url: Url,
+    issuer_url: Url,
+    client_id: String,
+    client_secret: Option<String>,
+) -> Result<CoreClient, Box<dyn Error>> {
+    let issuer_url = IssuerUrl::from_url(issuer_url);
+    let provider_metadata = CoreProviderMetadata::discover(&issuer_url, http_client)?;
+    let client = CoreClient::from_provider_metadata(
+        provider_metadata,
+        ClientId::new(client_id),
+        client_secret.map(ClientSecret::new),
+    )
+    .set_redirect_uri(RedirectUrl::from_url(redirect_url));
+    Ok(client)
+}
+
+#[repr(transparent)]
+#[derive(Clone, Debug)]
+pub struct Claims(CoreUserInfoClaims);
+
+impl Deref for Claims {
+    type Target = CoreUserInfoClaims;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+#[async_trait]
+impl<B: Send> FromRequest<B> for Claims {
+    type Rejection = Response;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        let redirect = Extension::<CoreClient>::from_request(req).await.unwrap();
+        let cookies = TypedHeader::<headers::Cookie>::from_request(req)
+            .await
+            .map_err(|e| {
+                error!("error when attempting to retrieve session token {e}");
```

Comment:
This should be `debug` at most, otherwise we'll get a spam of error logs on each invalid request.
Error logs are for critical server errors.